rat/codegen_backends.py

```diff
@@ -0,0 +1,302 @@
+        subscript_key = self.symbol_table.get_subscript_key(
+            self.primary_variable_name, self.primary_variable_subscript_names, variable_name, subscript_names, subscript_shifts
+        )
+        self.expression_string += f"subscripts['{subscript_key}']"
+
+    def generate(self, ast_node: ast.Expr, *args):
+        match ast_node:
+            case ast.Param():
+                self.expression_string += f"parameters['{ast_node.name}']"
+                if ast_node.subscript:
+                    if self.symbol_table.lookup(ast_node.name).pad_needed:
+                        # if padded: remove the zero-pad
+                        self.expression_string += "[:-1]"
+
+                    self.expression_string += "["
+                    self.generate(ast_node.subscript, ast_node.name)
+                    self.expression_string += "]"
+            case ast.Subscript():
+                self.generate_subscript(ast_node, *args)
+            case _:
+                super().generate(ast_node)
+
+
+class TransformedParametersCodeGenerator(EvaluateDensityCodeGenerator):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.lhs_used_in_rhs = False
+        self.lhs_key = ""
+        self.at_rhs = False
+
+    def generate(self, ast_node: ast.Expr, *args):
+        match ast_node:
+            case ast.Subscript():
+                super().generate(ast_node, *args)
+                if self.at_rhs:
+                    self.expression_string += "[index]"
+            case ast.Param():
+                param_key = ast_node.get_key()
+
+                if ast_node.subscript:
+                    if param_key == self.lhs_key:
+                        if self.at_rhs:
+                            # scan function
+                            subscript_node = ast_node.subscript
+                            subscript_names = tuple([x.name for x in subscript_node.names])
+                            subscript_shifts = tuple([x.value for x in subscript_node.shifts])
+                            subscript_key = self.symbol_table.get_subscript_key(
+                                self.primary_variable_name,
+                                self.primary_variable_subscript_names,
+                                param_key,
+                                subscript_names,
+                                subscript_shifts,
+                            )
+
+                            shift = [x for x in subscript_shifts if x != 0][0]
+                            self.expression_string += f"carry{shift}"
+                        else:
+                            self.expression_string += f"parameters['{param_key}']"
+                    elif param_key != self.lhs_key:
+                        self.expression_string += f"parameters['{param_key}']"
+                        self.expression_string += "["
+                        self.generate(ast_node.subscript, param_key)
```

Comment:
I pass param_key as an argument instead of using the member variable. I'm not sure if it is that much of an improvement because I didn't change is_rhs.

Reply:
Also the cleanup looks great.

Reply:
Hmm yeah I kept it as member variable to avoid `*args` hell but either seems fine.

Reply:
Okay I think the rule in a situation like this is solve the problem one way. I think that is_rhs is better implemented in the state. I'll revert this change.

Reply:
Reverted